File: medium/Hash-Table/valid-anagram/is_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        """
        Determine if t is an anagram of s.
        Approach: Hash map for character frequency counting
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if len(s) != len(t):
            return False

        count = {}

        # Count characters in s
        for c in s:
            count[c] = count.get(c, 0) + 1

        # Decrement count using t
        for c in t:
            count[c] = count.get(c, 0) - 1
            if count[c] < 0:
                return False

        return True

File: medium/Hash-Table/valid-anagram/is_anagram.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        /*
         * Check if t is an anagram of s.
         * Approach: Hash map frequency count
         * Time Complexity: O(n)
         * Space Complexity: O(1)
         */
        if (s.size() != t.size())
            return false;

        unordered_map<char, int> cnt;

        for (char c : s)
            cnt[c]++;

        for (char c : t) {
            if (--cnt[c] < 0)
                return false;
        }

        return true;
    }
};
